swarm/src/handler.rs

```diff
@@ -207,6 +207,8 @@ pub enum ConnectionEvent<'a, IP: InboundUpgradeSend, OP: OutboundUpgradeSend, IO
     AddressChange(AddressChange<'a>),
     /// Informs the handler that upgrading an outbound substream to the given protocol has failed.
     DialUpgradeError(DialUpgradeError<OOI, OP>),
+    /// Informs the handler that upgrading an outbound substream to the given protocol has expired its timeout.
+    DialTimeout(DialTimeout<OOI>),
```

Comment:
Would it make sense for `DialUpgradeError` to be an enum with a `Timeout` and `Upgrade` variant?

Reply:
https://github.com/libp2p/rust-libp2p/issues/3268 is related to this. If we agree to remove the `OpenInfo` from `ConnectionHandler` a lot of this code would be simpler because we wouldn't need to track `info` here.

Reply:
Off the top of my head this sounds good to me. I will have to give this more thought with a clear mind to give a good answer.